hey @Spagero763, looks like the file removal went too far. backend CI is now failing with:

```
src/controllers/stream.controller.ts(12,8): error TS2307: Cannot find module '../services/sorobanService.js'
src/routes/v1/stream.routes.ts(13,37): error TS2307: Cannot find module '../../controllers/stream/cancel.js'
```

`sorobanService.ts` and `controllers/stream/cancel.ts` are imported by other files on main, so they can't be fully deleted.

what i'd suggest:
- if your branch had **modifications** to those two files, do `git checkout origin/main -- backend/src/services/sorobanService.ts backend/src/controllers/stream/cancel.ts` to restore main's versions, then commit and push.
- the goal is for your final diff to only show your **frontend** test additions (and the `frontend/tsconfig.json` change), with no backend file modifications or deletions.

once the diff is just frontend files, CI will pass and we can merge.

--- a/frontend/src/__tests__/components.test.tsx
+++ b/frontend/src/__tests__/components.test.tsx
@@ -5,6 +5,9 @@ import React from 'react';
 // ─── LiveCounter ──────────────────────────────────────────────────────────────
 
 vi.mock('next/navigation', () => ({ useRouter: vi.fn() }));
+vi.mock('react-hot-toast', () => ({
+  default: { success: vi.fn(), error: vi.fn(), loading: vi.fn() },
+}));
 
 // Import after vi.mock registrations
 import LiveCounter from '../components/Livecounter';
@@ -41,7 +44,7 @@ describe('LiveCounter', () => {
     expect(label).not.toBeInTheDocument();
   });
 
-  it('resets to initial when isPaused switches to true', () => {
+  it('hides the streamed counter when isPaused becomes true', () => {
     const { rerender } = render(<LiveCounter initial={5} label="Streamed" />);
     act(() => { vi.advanceTimersByTime(3000); });
     rerender(<LiveCounter initial={5} label="Streamed" isPaused />);

--- a/frontend/src/__tests__/utils.test.ts
+++ b/frontend/src/__tests__/utils.test.ts
@@ -1,6 +1,16 @@
 import { describe, it, expect } from 'vitest';
+import { describe, it, expect, beforeEach } from 'vitest';
 import { convertArrayToCSV } from '../utils/csvExport';
 import { isValidStellarPublicKey } from '../lib/stellar';
+import {
+  fromStroops,
+  toStroops,
+  formatStreamRate,
+  hasValidPrecision,
+  getCachedTokenDecimals,
+  setCachedTokenDecimals,
+  clearTokenDecimalsCache,
+} from '../utils/amount';
 
 // ─── Amount / formatting utilities ───────────────────────────────────────────
 // The app stores raw i128 values (stroops) as strings; the dashboard divides
@@ -18,13 +28,6 @@ function parseAmount(tokenUnits: number): string {
 
 function formatRate(rawPerSecond: string): number {
   return parseFloat(rawPerSecond) / STROOPS_DIVISOR;
-}
-
-function hasValidPrecision(value: string, maxDecimals = 7): boolean {
-  if (!value || value.trim() === '') return false;
-  const parts = value.split('.');
-  if (parts.length === 1) return true;
-  return (parts[1]?.length ?? 0) <= maxDecimals;
 }
 
 describe('formatAmount', () => {
@@ -74,36 +77,9 @@ describe('formatRate', () => {
   });
 });
 
-describe('hasValidPrecision', () => {
-  it('accepts whole numbers', () => {
-    expect(hasValidPrecision('100')).toBe(true);
-    expect(hasValidPrecision('0')).toBe(true);
-  });
-
-  it('accepts values within the default 7-decimal limit', () => {
-    expect(hasValidPrecision('1.234')).toBe(true);
-    expect(hasValidPrecision('1.1234567')).toBe(true);
-  });
-
-  it('rejects values exceeding the 7-decimal limit', () => {
-    expect(hasValidPrecision('1.12345678')).toBe(false);
-  });
-
-  it('respects a custom maxDecimals argument', () => {
-    expect(hasValidPrecision('1.12', 2)).toBe(true);
-    expect(hasValidPrecision('1.123', 2)).toBe(false);
-  });
-
-  it('returns false for empty or whitespace strings', () => {
-    expect(hasValidPrecision('')).toBe(false);
-    expect(hasValidPrecision('   ')).toBe(false);
-  });
-});
-
 // ─── isValidStellarPublicKey ──────────────────────────────────────────────────
 
 describe('isValidStellarPublicKey (recipient validation)', () => {
-
   it('accepts a valid G-prefixed Ed25519 public key', () => {
     // Use a real randomly-generated testnet key
     const key = 'GDQERNIEDLE6SCKEAPO3ULKK5QQKFM3UIJMJQNBMKXPQR6HDYQTM2WO';
@@ -171,5 +147,126 @@ describe('convertArrayToCSV', () => {
   it('handles null and undefined cell values as empty strings', () => {
     const csv = convertArrayToCSV([{ a: null, b: undefined, c: 'ok' }]);
     expect(csv.split('\n')[1]).toBe(',,ok');
+  });
+});
+
+// ─── amount.ts ────────────────────────────────────────────────────────────────
+
+describe('fromStroops', () => {
+  it('converts 0 decimals — returns raw string', () => {
+    expect(fromStroops(42n, 0)).toBe('42');
+  });
+
+  it('converts to exact token units', () => {
+    expect(fromStroops(10_000_000n, 7)).toBe('1');
+    expect(fromStroops(50_000_000n, 7)).toBe('5');
+  });
+
+  it('produces fractional output', () => {
+    expect(fromStroops(5_000_000n, 7)).toBe('0.5');
+    expect(fromStroops(1n, 7)).toBe('0.0000001');
+  });
+
+  it('trims trailing zeros in fractional part', () => {
+    expect(fromStroops(12_300_000n, 7)).toBe('1.23');
+  });
+
+  it('handles zero', () => {
+    expect(fromStroops(0n, 7)).toBe('0');
+  });
+
+  it('handles large amounts', () => {
+    expect(fromStroops(1_000_000_000_000n, 7)).toBe('100000');
+  });
+
+  it('round-trips with toStroops', () => {
+    expect(fromStroops(toStroops('3.14', 7), 7)).toBe('3.14');
+  });
+});
+
+describe('toStroops', () => {
+  it('converts whole token units to stroops', () => {
+    expect(toStroops('1', 7)).toBe(10_000_000n);
+    expect(toStroops('5', 7)).toBe(50_000_000n);
+  });
+
+  it('converts fractional token units', () => {
+    expect(toStroops('0.5', 7)).toBe(5_000_000n);
+    expect(toStroops('0.0000001', 7)).toBe(1n);
+  });
+
+  it('returns 0n for empty/whitespace string', () => {
+    expect(toStroops('', 7)).toBe(0n);
+    expect(toStroops('   ', 7)).toBe(0n);
+  });
+
+  it('truncates fractional part exceeding decimals', () => {
+    expect(toStroops('1.12345678', 7)).toBe(toStroops('1.1234567', 7));
+  });
+
+  it('pads short fractional parts', () => {
+    expect(toStroops('1.5', 7)).toBe(15_000_000n);
+  });
+
+  it('handles 0 decimals', () => {
+    expect(toStroops('42', 0)).toBe(42n);
+  });
+});
+
+describe('formatStreamRate', () => {
+  it('includes per-second and per-month rates', () => {
+    const result = formatStreamRate(10_000_000n, 7, 'USDC');
+    expect(result).toContain('USDC/sec');
+    expect(result).toContain('USDC/month');
+  });
+
+  it('uses USDC as the default token symbol', () => {
+    expect(formatStreamRate(10_000_000n, 7)).toContain('USDC');
+  });
+
+  it('shows 0 rate as "0"', () => {
+    expect(formatStreamRate(0n, 7, 'XLM')).toMatch(/^0 XLM\/sec/);
+  });
+});
+
+describe('hasValidPrecision (amount.ts)', () => {
+  it('returns true for empty string', () => {
+    expect(hasValidPrecision('', 7)).toBe(true);
+  });
+
+  it('accepts whole numbers', () => {
+    expect(hasValidPrecision('100', 7)).toBe(true);
+  });
+
+  it('accepts fractional values within limit', () => {
+    expect(hasValidPrecision('1.1234567', 7)).toBe(true);
+  });
+
+  it('rejects values exceeding the decimal limit', () => {
+    expect(hasValidPrecision('1.12345678', 7)).toBe(false);
+  });
+
+  it('respects a custom decimal limit', () => {
+    expect(hasValidPrecision('1.12', 2)).toBe(true);
+    expect(hasValidPrecision('1.123', 2)).toBe(false);
+  });
+});
+
+describe('token decimals cache', () => {
+  beforeEach(() => { clearTokenDecimalsCache(); });
+
+  it('returns undefined for uncached address', () => {
+    expect(getCachedTokenDecimals('GABC')).toBeUndefined();
+  });
+
+  it('stores and retrieves decimals', () => {
+    setCachedTokenDecimals('GTOKEN', 7);
+    expect(getCachedTokenDecimals('GTOKEN')).toBe(7);
+  });
+
+  it('clearTokenDecimalsCache wipes all entries', () => {
+    setCachedTokenDecimals('GTOKEN', 7);
+    clearTokenDecimalsCache();
+    expect(getCachedTokenDecimals('GTOKEN')).toBeUndefined();
   });
 });

--- a/frontend/src/__tests__/utils.test.ts
+++ b/frontend/src/__tests__/utils.test.ts
@@ -1,111 +1,94 @@
 import { describe, it, expect } from 'vitest';
-import { describe, it, expect, beforeEach } from 'vitest';
 import { convertArrayToCSV } from '../utils/csvExport';
 import { isValidStellarPublicKey } from '../lib/stellar';
 import {
-  fromStroops,
-  toStroops,
-  formatStreamRate,
+  formatAmount,
+  parseAmount,
+  formatRate,
   hasValidPrecision,
-  getCachedTokenDecimals,
-  setCachedTokenDecimals,
-  clearTokenDecimalsCache,
-} from '../utils/amount';
-
-// ─── Amount / formatting utilities ───────────────────────────────────────────
-// The app stores raw i128 values (stroops) as strings; the dashboard divides
-// by 1e7 to convert to token units.  We test that conversion arithmetic here.
-
-const STROOPS_DIVISOR = 1e7;
-
-function formatAmount(raw: string): number {
-  return parseFloat(raw) / STROOPS_DIVISOR;
-}
-
-function parseAmount(tokenUnits: number): string {
-  return Math.round(tokenUnits * STROOPS_DIVISOR).toString();
-}
-
-function formatRate(rawPerSecond: string): number {
-  return parseFloat(rawPerSecond) / STROOPS_DIVISOR;
-}
+  toStroops,
+  fromStroops,
+  truncateAmount,
+  formatCompactAmount,
+} from '../lib/amount';
 
 describe('formatAmount', () => {
-  it('converts raw i128 stroops to token units', () => {
-    expect(formatAmount('10000000')).toBe(1);
-    expect(formatAmount('50000000')).toBe(5);
-    expect(formatAmount('0')).toBe(0);
+  it('converts raw bigint amounts to token units', () => {
+    expect(formatAmount(10_000_000n, 7)).toBe('1');
+    expect(formatAmount(50_000_000n, 7)).toBe('5');
+    expect(formatAmount(0n, 7)).toBe('0');
   });
 
-  it('handles fractional results', () => {
-    expect(formatAmount('5000000')).toBeCloseTo(0.5);
-    expect(formatAmount('1')).toBeCloseTo(1e-7);
-  });
-
-  it('handles large amounts', () => {
-    expect(formatAmount('1000000000000')).toBeCloseTo(100000);
+  it('preserves fractional precision and trims trailing zeros', () => {
+    expect(formatAmount(5_000_000n, 7)).toBe('0.5');
+    expect(formatAmount(1n, 7)).toBe('0.0000001');
+    expect(formatAmount(12_300_000n, 7)).toBe('1.23');
   });
 });
 
 describe('parseAmount', () => {
-  it('converts token units back to raw i128 string', () => {
-    expect(parseAmount(1)).toBe('10000000');
-    expect(parseAmount(5)).toBe('50000000');
-    expect(parseAmount(0)).toBe('0');
+  it('converts token strings back to raw bigint amounts', () => {
+    expect(parseAmount('1', 7)).toBe(10_000_000n);
+    expect(parseAmount('5', 7)).toBe(50_000_000n);
+    expect(parseAmount('0', 7)).toBe(0n);
   });
 
   it('round-trips correctly', () => {
-    const original = '12345000';
-    expect(parseAmount(formatAmount(original))).toBe(original);
+    const original = '123.45';
+    expect(parseAmount(formatAmount(parseAmount(original, 7), 7), 7)).toBe(parseAmount(original, 7));
   });
 
-  it('rounds to the nearest stroop', () => {
-    // 0.12345678 XLM → rounds at 7 decimal places
-    const result = parseAmount(0.1234567);
-    expect(parseInt(result, 10)).toBeGreaterThan(0);
+  it('pads and truncates fractional input as expected', () => {
+    expect(parseAmount('1.5', 7)).toBe(15_000_000n);
+    expect(parseAmount('1.12345678', 7)).toBe(parseAmount('1.1234567', 7));
+    expect(parseAmount('', 7)).toBe(0n);
   });
 });
 
 describe('formatRate', () => {
-  it('converts raw rate per second to token units per second', () => {
-    expect(formatRate('10000000')).toBe(1); // 1 token/sec
-    expect(formatRate('100')).toBeCloseTo(0.00001);
+  it('converts a raw per-second rate to a readable string', () => {
+    expect(formatRate(10_000_000n, 7, 'USDC')).toBe('1 USDC/sec (86400 USDC/day)');
   });
 
   it('returns 0 for a zero rate', () => {
-    expect(formatRate('0')).toBe(0);
+    expect(formatRate(0n, 7, 'XLM')).toBe('0');
   });
 });
 
-// ─── isValidStellarPublicKey ──────────────────────────────────────────────────
+describe('hasValidPrecision', () => {
+  it('accepts whole numbers and empty input', () => {
+    expect(hasValidPrecision('', 7)).toBe(true);
+    expect(hasValidPrecision('100', 7)).toBe(true);
+    expect(hasValidPrecision('0', 7)).toBe(true);
+  });
+
+  it('accepts values within the decimal limit', () => {
+    expect(hasValidPrecision('1.234', 7)).toBe(true);
+    expect(hasValidPrecision('1.1234567', 7)).toBe(true);
+  });
+
+  it('rejects values exceeding the decimal limit', () => {
+    expect(hasValidPrecision('1.12345678', 7)).toBe(false);
+  });
+
+  it('respects a custom decimal limit', () => {
+    expect(hasValidPrecision('1.12', 2)).toBe(true);
+    expect(hasValidPrecision('1.123', 2)).toBe(false);
+  });
+});
 
 describe('isValidStellarPublicKey (recipient validation)', () => {
   it('accepts a valid G-prefixed Ed25519 public key', () => {
-    // Use a real randomly-generated testnet key
     const key = 'GDQERNIEDLE6SCKEAPO3ULKK5QQKFM3UIJMJQNBMKXPQR6HDYQTM2WO';
-    // StrKey validation requires the correct checksum — test with known valid keys
     expect(typeof isValidStellarPublicKey(key)).toBe('boolean');
   });
 
-  it('rejects an empty string', () => {
+  it('rejects empty, short, and wrong-prefix values', () => {
     expect(isValidStellarPublicKey('')).toBe(false);
-  });
-
-  it('rejects a string that is too short', () => {
     expect(isValidStellarPublicKey('GABC123')).toBe(false);
-  });
-
-  it('rejects a key with a wrong prefix', () => {
     expect(isValidStellarPublicKey('SABC123XYZ456DEF789GHI012JKL345MNO678PQR901STU234VWX567YZA')).toBe(false);
   });
-
-  it('trims surrounding whitespace before validating', () => {
-    // isValidStellarPublicKey normalises the input
-    expect(isValidStellarPublicKey('  ')).toBe(false);
-  });
 });
-
-// ─── CSV export utilities ─────────────────────────────────────────────────────
 
 describe('convertArrayToCSV', () => {
   it('returns empty string for null/undefined input', () => {
@@ -129,18 +112,14 @@ describe('convertArrayToCSV', () => {
     ];
     const csv = convertArrayToCSV(rows);
     const lines = csv.split('\n');
-    expect(lines).toHaveLength(3); // header + 2 data rows
+    expect(lines).toHaveLength(3);
     expect(lines[1]).toBe('1,hello');
     expect(lines[2]).toBe('2,world');
   });
 
-  it('escapes cells that contain commas', () => {
-    const csv = convertArrayToCSV([{ name: 'Doe, Jane', value: '5' }]);
+  it('escapes cells that contain commas and quotes', () => {
+    const csv = convertArrayToCSV([{ name: 'Doe, Jane', note: 'say "hello"', value: '5' }]);
     expect(csv).toContain('"Doe, Jane"');
-  });
-
-  it('escapes cells that contain double-quotes', () => {
-    const csv = convertArrayToCSV([{ note: 'say "hello"', v: '1' }]);
     expect(csv).toContain('""hello""');
   });
 
@@ -150,123 +129,25 @@ describe('convertArrayToCSV', () => {
   });
 });
 
-// ─── amount.ts ────────────────────────────────────────────────────────────────
-
-describe('fromStroops', () => {
-  it('converts 0 decimals — returns raw string', () => {
-    expect(fromStroops(42n, 0)).toBe('42');
-  });
-
-  it('converts to exact token units', () => {
-    expect(fromStroops(10_000_000n, 7)).toBe('1');
-    expect(fromStroops(50_000_000n, 7)).toBe('5');
-  });
-
-  it('produces fractional output', () => {
-    expect(fromStroops(5_000_000n, 7)).toBe('0.5');
-    expect(fromStroops(1n, 7)).toBe('0.0000001');
-  });
-
-  it('trims trailing zeros in fractional part', () => {
-    expect(fromStroops(12_300_000n, 7)).toBe('1.23');
-  });
-
-  it('handles zero', () => {
-    expect(fromStroops(0n, 7)).toBe('0');
-  });
-
-  it('handles large amounts', () => {
-    expect(fromStroops(1_000_000_000_000n, 7)).toBe('100000');
-  });
-
-  it('round-trips with toStroops', () => {
-    expect(fromStroops(toStroops('3.14', 7), 7)).toBe('3.14');
+describe('toStroops and fromStroops', () => {
+  it('converts between display strings and stroops using 7 decimals', () => {
+    expect(toStroops('1')).toBe(10_000_000n);
+    expect(toStroops('0.5')).toBe(5_000_000n);
+    expect(fromStroops(10_000_000n)).toBe('1');
+    expect(fromStroops(42n)).toBe('0.0000042');
   });
 });
 
-describe('toStroops', () => {
-  it('converts whole token units to stroops', () => {
-    expect(toStroops('1', 7)).toBe(10_000_000n);
-    expect(toStroops('5', 7)).toBe(50_000_000n);
-  });
-
-  it('converts fractional token units', () => {
-    expect(toStroops('0.5', 7)).toBe(5_000_000n);
-    expect(toStroops('0.0000001', 7)).toBe(1n);
-  });
-
-  it('returns 0n for empty/whitespace string', () => {
-    expect(toStroops('', 7)).toBe(0n);
-    expect(toStroops('   ', 7)).toBe(0n);
-  });
-
-  it('truncates fractional part exceeding decimals', () => {
-    expect(toStroops('1.12345678', 7)).toBe(toStroops('1.1234567', 7));
-  });
-
-  it('pads short fractional parts', () => {
-    expect(toStroops('1.5', 7)).toBe(15_000_000n);
-  });
-
-  it('handles 0 decimals', () => {
-    expect(toStroops('42', 0)).toBe(42n);
+describe('truncateAmount', () => {
+  it('truncates without rounding', () => {
+    expect(truncateAmount(12_345_678_900n, 7, 3)).toBe('1234.567');
+    expect(truncateAmount(0n, 7, 3)).toBe('0');
   });
 });
 
-describe('formatStreamRate', () => {
-  it('includes per-second and per-month rates', () => {
-    const result = formatStreamRate(10_000_000n, 7, 'USDC');
-    expect(result).toContain('USDC/sec');
-    expect(result).toContain('USDC/month');
-  });
-
-  it('uses USDC as the default token symbol', () => {
-    expect(formatStreamRate(10_000_000n, 7)).toContain('USDC');
-  });
-
-  it('shows 0 rate as "0"', () => {
-    expect(formatStreamRate(0n, 7, 'XLM')).toMatch(/^0 XLM\/sec/);
-  });
-});
-
-describe('hasValidPrecision (amount.ts)', () => {
-  it('returns true for empty string', () => {
-    expect(hasValidPrecision('', 7)).toBe(true);
-  });
-
-  it('accepts whole numbers', () => {
-    expect(hasValidPrecision('100', 7)).toBe(true);
-  });
-
-  it('accepts fractional values within limit', () => {
-    expect(hasValidPrecision('1.1234567', 7)).toBe(true);
-  });
-
-  it('rejects values exceeding the decimal limit', () => {
-    expect(hasValidPrecision('1.12345678', 7)).toBe(false);
-  });
-
-  it('respects a custom decimal limit', () => {
-    expect(hasValidPrecision('1.12', 2)).toBe(true);
-    expect(hasValidPrecision('1.123', 2)).toBe(false);
-  });
-});
-
-describe('token decimals cache', () => {
-  beforeEach(() => { clearTokenDecimalsCache(); });
-
-  it('returns undefined for uncached address', () => {
-    expect(getCachedTokenDecimals('GABC')).toBeUndefined();
-  });
-
-  it('stores and retrieves decimals', () => {
-    setCachedTokenDecimals('GTOKEN', 7);
-    expect(getCachedTokenDecimals('GTOKEN')).toBe(7);
-  });
-
-  it('clearTokenDecimalsCache wipes all entries', () => {
-    setCachedTokenDecimals('GTOKEN', 7);
-    clearTokenDecimalsCache();
-    expect(getCachedTokenDecimals('GTOKEN')).toBeUndefined();
+describe('formatCompactAmount', () => {
+  it('formats large amounts with compact notation', () => {
+    expect(formatCompactAmount(10_000_000_000n, 7)).toBe('1.0K');
+    expect(formatCompactAmount(2_500_000_000_000n, 7)).toBe('250.0K');
   });
 });

--- a/frontend/tsconfig.json
+++ b/frontend/tsconfig.json
@@ -30,5 +30,5 @@
     ".next/dev/types/**/*.ts",
     "**/*.mts"
   ],
-  "exclude": ["node_modules"]
+  "exclude": ["node_modules", "vitest.config.ts"]
 }

